File: app01_java_block/src/main/java/com/devpowers/legalchain/api/TransactionController.java
package com.devpowers.legalchain.api;

import com.devpowers.legalchain.core.Blockchain;
import com.devpowers.legalchain.core.NodeWallet;
import com.devpowers.legalchain.core.Transaction;
import com.devpowers.legalchain.core.TransactionType;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.PositiveOrZero;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Token transfers. In the MVP the browser has no key store, so the node wallet signs
 * on the user's behalf (custodial model); the ML-DSA signature is created here and
 * verified again by {@link Blockchain#submit} — the ledger never trusts the API layer.
 */
@RestController
@RequestMapping("/api/transactions")
public class TransactionController {

    private final Blockchain blockchain;
    private final NodeWallet wallet;

    public TransactionController(Blockchain blockchain, NodeWallet wallet) {
        this.blockchain = blockchain;
        this.wallet = wallet;
    }

    @GetMapping("/pending")
    public List<Transaction> pending() {
        return blockchain.pendingTransactions();
    }

    @PostMapping
    public Transaction submit(@Valid @RequestBody TransferRequest request) {
        Transaction unsigned = new Transaction(
                UUID.randomUUID().toString(),
                System.currentTimeMillis(),
                wallet.address(),
                request.recipient(),
                request.amount(),
                TransactionType.TRANSFER,
                request.memo() == null ? Map.of() : Map.of("memo", request.memo()),
                wallet.publicKeyBase64(),
                null);
        Transaction signed = new Transaction(
                unsigned.id(), unsigned.timestamp(), unsigned.sender(), unsigned.recipient(),
                unsigned.amount(), unsigned.type(), unsigned.payload(),
                unsigned.senderPublicKey(), wallet.sign(unsigned.contentToSign()));
        return blockchain.submit(signed);
    }

    public record TransferRequest(
            @NotBlank String recipient,
            @NotNull @PositiveOrZero BigDecimal amount,
            String memo) {
    }
}

File: app03_csharp_block/src/LegalChain/Api/TransactionController.cs
using System.ComponentModel.DataAnnotations;
using LegalChain.Core;
using Microsoft.AspNetCore.Mvc;

namespace LegalChain.Api;

/// <summary>
/// Token transfers. In the MVP the browser has no key store, so the node wallet signs
/// on the user's behalf (custodial model); the ML-DSA signature is created here and
/// verified again by <see cref="Blockchain.Submit"/> — the ledger never trusts the API layer.
/// </summary>
[ApiController]
[Route("/api/transactions")]
public class TransactionController : ControllerBase
{
    private readonly Blockchain _blockchain;
    private readonly NodeWallet _wallet;

    public TransactionController(Blockchain blockchain, NodeWallet wallet)
    {
        _blockchain = blockchain;
        _wallet = wallet;
    }

    [HttpGet("pending")]
    public IReadOnlyList<Transaction> Pending() => _blockchain.PendingTransactions();

    [HttpPost]
    public Transaction Submit([FromBody] TransferRequest request)
    {
        var unsigned = new Transaction(
            Guid.NewGuid().ToString(),
            DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            _wallet.Address,
            request.Recipient,
            request.Amount,
            TransactionType.TRANSFER,
            request.Memo is null
                ? new Dictionary<string, string>()
                : new Dictionary<string, string> { ["memo"] = request.Memo },
            _wallet.PublicKeyBase64,
            null);
        Transaction signed = unsigned with { Signature = _wallet.Sign(unsigned.ContentToSign()) };
        return _blockchain.Submit(signed);
    }

    public record TransferRequest(
        [property: Required(AllowEmptyStrings = false)] string Recipient,
        [property: Range(0, double.MaxValue)] decimal Amount,
        string? Memo);
}
